languages/swift/exercises/concept/higher-order-functions/.docs/introduction.md

```diff
@@ -0,0 +1,120 @@
+
+```swift
+var stringAndIntToString: (String, Int) -> String
+
+stringAndIntToString = shoppingList
+stringAndIntToString("carrots", 3)
+// => "You need to buy 3 carrots."
+
+stringAndIntToString = repeater
+stringAndIntToString("carrots", 3)
+// => "carrotscarrotscarrots"
+```
+
+As with other variables/constants, the value you assign to it must be of the correct type:
+
+```swift
+stringAndIntToString = printShoppingList
+// Error: Cannot assign value of type '(String) -> ()' to type '(String, Int) -> String'
+```
+
+### Function types as parameter types
+
+As function types can be used anywhere other types can be used, it follows that they can be used as parameter types and passed into other functions and called from within those functions.
+
+```swift
+func distribute(over strings: [String],
+                andInt i: Int,
+                function f: (String, Int) -> String
+               ) -> [String] {
+  var result = [String]()
+  for string in strings {
+    result.append(f(string, i))
+  }
+  return result
+}
+// distribute: ([String], Int, (String, Int) -> String) -> [String]
+
+distribute(over: ["carrots", "eggs", "popsicles"], andInt: 3, function: shoppingList)
+// => ["You need to buy 3 carrots.", "You need to buy 3 eggs.", "You need to buy 3 popsicles."]
+
+distribute(over: ["carrots", "eggs", "popsicles"], andInt: 3, function: repeater)
+// => ["carrotscarrotscarrots", "eggseggseggs", "popsiclespopsiclespopsicles"]
+
+func apply(functions fs: [(String, Int) -> String],
+           to string: String,
+           andInt i: Iny
+          ) -> [String] {
+  var result = [String]()
+  for f in fs {
+    result.append(f(string, i))
+  }
+  return result
+
+}
+// apply: ([(String, Int) -> String], String, Int) -> [String]
+
+
+apply(functions: [shoppingList, repeater], to: "carrots", andInt: 3)
+// => ["You need to buy 3 carrots.", "carrotscarrotscarrots"]
+```
+
+### Function types as return types
+
+Similarly, function types may be used as return types for functions. In other words, one can write functions that create and return other functions. When creating these functions, the function that is returned can use the other parameters passed into the function or the local variables created inside the parent function.
+
+```swift
+func makeAdder(base: Int) -> (Int) -> Int {
+  func adder(_ i: Int) -> Int {
+    base + i
+  }
+  return adder
+}
+// makeAdder: (Int) -> (Int) -> Int
+
+let add10 = makeAdder(base: 10)
+// add10: (Int) -> Int
+
+let subtract20 = makeAdder(base: -20)
+// subtract10: (Int) -> Int
+
+add10(5)
+// => 15
+
+subtract20(5)
+// => -15
+```
+
+### Higher-order functions
+
+Functions that take functions as parameters and/or return functions are known as higher order functions. Swift has many higher-order functions in its various libraries.
```

Comment:
Does the student have to know this to solve the exercise? If not, I would suggest removing it here. It is definitely great for the `after.md` document.